### PR TITLE
pull request so you could check it

### PR DESCRIPTION
**Pull request by Elyamira**

### CODE REVIEW

File: pages/index.tsx [hunk 1]
```diff
@@ -1,72 +1,88 @@
...
-        </p>
+  useEffect(() => {
+    if (errorWhileCatchingShows) return;
+    if (!countryCodeError && countryCode && schedule.length < 1) {
+      getAllShows(`${Endpoints.tvScheduleEndpoint}?country=${countryCode}`);
+    }
+    if (countryCodeError) {
+      getAllShows(Endpoints.tvScheduleEndpoint);
+    }
+  })
```
Comment (chiubaca) [inline pages/index.tsx]: You usually will want to ensure you have populated your dependency array with the props it's consuming otherwise this effect may re-execute on every render unnecessarily. 

```suggestion
  useEffect(() => {
    if (errorWhileCatchingShows) return;
    if (!countryCodeError && countryCode && schedule.length < 1) {
      getAllShows(`${Endpoints.tvScheduleEndpoint}?country=${countryCode}`);
    }
    if (countryCodeError) {
      getAllShows(Endpoints.tvScheduleEndpoint);
    }
  }, [countryCode, countryCodeError, errorWhileCatchingShows, schedule.length]);
```

File: src/presentation/pages/home/index.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+const HomePageLayout = (props: { schedule: Schedule[]; }) => {
+    const { schedule } = props;
+    const [currentPage, setCurrentPage] = useState(1);
+    // No of Items to be displayed on each page   
+    const [itemsPerPage] = useState(12);
+    const indexOfLastPageItem = currentPage * itemsPerPage;
+    const indexOfFirstPageItem = indexOfLastPageItem - itemsPerPage;
+    const currentPageItems = schedule?.slice(indexOfFirstPageItem, indexOfLastPageItem);
+    const nPages = Math.ceil(schedule.length / itemsPerPage);
```
Comment (chiubaca) [inline src/presentation/pages/home/index.tsx]: opportunity to encapsulate this logic into a [react hook](https://reactjs.org/docs/hooks-custom.html#extracting-a-custom-hook). You have similar logic in `src/presentation/pages/single-show/SingleShowPageLayoutStyled.ts`

File: src/presentation/pages/single-show/index.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                                        }
+                                        alt="raiting star"
+                                        width={24}
+                                        height={24}
+                                    />
+                                );
+                            })}
+                        </div>
+                        <p>{`${rating}/5`}</p>
+                    </div>
```
Comment (chiubaca) [inline src/presentation/pages/single-show/index.tsx]: Opportunity to convert this into `StarRating` component as you've got the same logic repeated in `src/presentation/elements/schedule-tv-show-card/index.tsx`

**Comment by chiubaca** [comment]
## Good Things
- Tests!
- Typescript!
	- Especially good job creating the types for the API 
- Semantic git commits!
- You’ve taken the time to organise your code in a sensible structure which is great to see.
	- I especially like how you separated small helper utilities into the services directory.

- Grouping constants and “magic numbers” into `src/config/Settings.ts` is a great pattern.
- Good work on the pagination logic.
- Good to see use of semantic HTML, but remember to take into consideration of [heading ordering](src/presentation/pages/single-show/index.tsx). Ensure your page starts with a `h1` and doesn’t skip headings for accessibility.

## Issues
 - There are a few UI issues:
	   - The footer is not fixed to the bottom.
	   - There is some layout shift when paginating between cast members.
	   - The top row cards on the home page appear at a lower z-index to your grey banner causing it to fall behind and look awkward.

- I ran into a bug that caused the landing page to continuously flash, I think it was because I have an extension that was blocking requests to https://geolocation-db.com/json/.  It looks like it was due to a missing dependency array in `/pages/index.tsx` causing an infinite loop, [see here](https://github.com/Elyamira/tv-maze/pull/1#discussion_r959709387).

- This is up for debate, but your use of classes doesn’t seem appropriate. As you are only using static methods, you’re better off exporting plain functions. The main benefit of this would be tree-shaking. E.g in the case that you only wanted to use `getStarsFromTvRating` but not `getStarsWithDecimals`, you would not need to import the entire class.

- Your usage of `styled-components` though works, is not ideal. A more idiomatic way of using styled-component would be to break down your UI so you can reason with your CSS in smaller blocks rather than colocating all your CSS code in one giant block. Let’s take a small example from `SingleShowPageLayout.tsx` as a quick example. (lots of code truncated for brevity).

```tsx
export const Example = () => {
	return (	
		<div className="starring-info-with-header">		
			<h3 className="bottom-part-header">Starring</h3>		
			<div className="starring-info">{/* omitted code for brevity */}</div>		
		</div>	
	);
};
```
could be refactored to look like this instead:

```tsx
const StarringInfoWithHeader = styled.div`
  /* Your custom css */
`;

const BottomPartHeader = styled.h3`
  /* Your custom css */
`;

const StaringInfo = styled.div`
  /* Your custom css */
`;

  

export const Example = () => {
	return (
		<StarringInfoWithHeader>
			<BottomPartHeader>Starring</BottomPartHeader>
			<StaringInfo>{/* omitted code for brevity */}</StaringInfo>
		</StarringInfoWithHeader>	
	);
};
```

## Closing comments

Overall Great job! Your project is extremely well organised which is great to see. The separation of logic into `config`, `domain` and `presentation` directories is different and not something I’ve seen before but very intuitive, I like it!

You get extra brownie points for choosing to use typescript and going the extra mile with typing the API schema. The types do mean extra work but you’ve demonstrated your understanding of type narrowing and guarded against undefined and null types.

Well done for adding tests though basic, it's definitely better than none.

Overall, the front end is quite basic but meets all the requirements. It feels like you have brute-forced your way through the css styling resulting in some quite scary css code. I would love to get your thoughts on how you could make this more maintainable and easier to read.

Well done! 💪🏼

**Comment by Elyamira** [comment]
Hey Alex! 
Thank you for such a detailed review. I agree with all the points you mentioned above (sorry for an infinite loop due to missing dependency array). I wrote a class just to show that I can write/ use classes in a project, but indeed in that scenario functions suit better.  Thanks for sharing an example of how CSS styles can be improved. 
I agree that now styles look a bit scary, and probably more components could be styled separately. I also think that writing media queries inside desktop styles was not the best solution. The styles could look more clear with media queries below all styles. To make styles look better I could also create some wrappers to style some code instead of repeating several patterns, e.g. wrapper with main paddings which I applied to every component.
Moreover, it would be good to create design system styles, with the project's colours, and add paddings and margins so all gaps between components are consistent and follow the same pattern which can be really helpful while adding something new to the project. 
As for tests, yes there are a few tests only, but I wanted to add them to show that I can write tests. I'm still learning how to write proper tests, though. 
I'm still learning how to write better and what technologies to use, what structure and so on, so your comments are really helpful for me to understand in which areas I lack some fundamental knowledge and to fill these gaps later. Thank you a lot!

File: pages/show/[showId]/index.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    }
+    else return <ErrorAlert />
+
+}
+export const getServerSideProps: GetServerSideProps = async (context) => {
+    const showId = context.params?.showId;
+    if (!showId) {
+        return {
+            redirect: {
+                destination: '/404',
```
Comment (MatrixAlain) [inline pages/show/[showId]/index.tsx:54]: This doesn't actually redirect to the in-built 404 page - quite funnily it ends up at the 404 page because it fails to resolve, so is a 404 error. :)

A better way to do this would be to:

`import Error from 'next/error'`

and then:

`return <Error statusCode={404} />`

You can see this in the [Next documentation here](https://nextjs.org/docs/advanced-features/custom-error-page#404-page)
